src/libopensc/card-edo.c

```diff
@@ -0,0 +1,332 @@
+ *
+ * If ECDSA was used, the ASN.1 sequence of integers R,S returned by the
+ * card needs to be converted to the raw concatenation of R,S for PKCS#11.
+ */
+static int edo_compute_signature(struct sc_card* card, const u8* data, size_t datalen, u8* out, size_t outlen) {
+	LOG_FUNC_CALLED(card->ctx);
+	u8 sig[SC_MAX_APDU_RESP_SIZE];
+	LOG_TEST_RET(card->ctx, sc_get_iso7816_driver()->ops->compute_signature(card, data, datalen, sig, sizeof sig), "Internal signature failed");
+	LOG_TEST_RET(card->ctx, sc_asn1_sig_value_sequence_to_rs(card->ctx, sig, sizeof sig, out, outlen), "ASN.1 conversion failed");
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Sets security environment
+ *
+ * Card expects key file to be selected first, followed by the
+ * set security env packet with: 0x80, 0x01, 0xcc, 0x84, 0x01, 0x80|x,
+ * where x is the key reference byte.
+ */
+static int edo_set_security_env(struct sc_card* card, const struct sc_security_env* env, int se_num) {
+	LOG_FUNC_CALLED(card->ctx);
+	struct sc_apdu apdu;
+
+	if (env->algorithm == SC_ALGORITHM_EC && env->operation == SC_SEC_OPERATION_SIGN && env->flags & SC_SEC_ENV_KEY_REF_PRESENT) {
+		u8 payload[] = {0x80, 0x01, 0xcc, 0x84, 0x01, 0x80 | env->key_ref[0]};
+		sc_format_apdu_ex(&apdu, 0x00, 0x22, 0x41, 0xB6, payload, sizeof payload, NULL, 0);
+	} else
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_NOT_SUPPORTED);
+
+	LOG_TEST_RET(card->ctx, sc_select_file(card, &env->file_ref, NULL), "SELECT file failed");
+	LOG_TEST_RET(card->ctx, sc_transmit_apdu(card, &apdu), "APDU transmit failed");
+	LOG_TEST_RET(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2), "SW check failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+/*! Initializes card driver.
+ *
+ * Card is known to support only short APDU-s.
+ * Preinitialized keys are on secp384r1 curve.
+ * PACE channel have to be established.
+ */
+static int edo_init(sc_card_t* card) {
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	edo_eac_init();
+
+	memset(&card->sm_ctx, 0, sizeof card->sm_ctx);
+
+	card->max_send_size = SC_MAX_APDU_RESP_SIZE;
+	card->max_recv_size = SC_MAX_APDU_RESP_SIZE;
+
+	for (size_t i = 0; i < sizeof edo_curves / sizeof * edo_curves; ++i) {
+		LOG_TEST_RET(card->ctx, _sc_card_add_ec_alg(
+			card, edo_curves[i].len,
+			SC_ALGORITHM_ECDSA_RAW | SC_ALGORITHM_ECDSA_HASH_NONE,
+			0, &edo_curves[i].oid
+		), "Add EC alg failed");
+	}
+
+	LOG_TEST_RET(card->ctx, sc_enum_apps(card), "Enumerate apps failed");
+
+	LOG_TEST_RET(card->ctx, edo_unlock(card), "Unlock card failed");
+
+	LOG_FUNC_RETURN(card->ctx, SC_SUCCESS);
+}
+
+
+struct sc_card_driver* sc_get_edo_driver(void) {
+	edo_ops = *sc_get_iso7816_driver()->ops;
+	edo_ops.match_card = edo_match_card;
+	edo_ops.init = edo_init;
+	edo_ops.select_file = edo_select_file;
+	edo_ops.set_security_env = edo_set_security_env;
+	edo_ops.compute_signature = edo_compute_signature;
+
+	return &edo_drv;
+}
+
+#else
+
+struct sc_card_driver* sc_get_edo_driver(void) {
+	return 0;
```

Comment:
SC_ERROR_NOT_SUPPORTED should be more appropriate

Reply:
The function is expected to return pointer, so `-1203` will not be suitable here. I have chosen to make this NULL stub, to avoid creating missing symbol which is declared as extern in `cards.h`. It's used later in the `ctx.c` when populating `internal_card_drivers` but only when openSSL and openPACE are enabled.

Reply:
I see, please use `NULL` then, thanks.